File: report_dashboard/charts.py
"""순수 SVG 생성기 — 목업 charts.js를 파이썬으로 옮긴 것. Streamlit에 의존하지 않는다.

스펙 §4.1: 선 2px, 끝점 r4.5(+표면색 2px 링은 CSS .end), 격자 헤어라인 3줄, 워시는 <g class="fade-late">로
감싸서 애니메이션이 path의 opacity 토큰을 덮지 않게 한다. 색은 전부 CSS 클래스(테마 토큰)로만 정한다.
"""
from __future__ import annotations

from html import escape


def path_length(points: list[tuple[float, float]]) -> float:
    total = 0.0
    for (x1, y1), (x2, y2) in zip(points, points[1:]):
        total += ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5
    return total


def _points(values, width, height, pad_l, pad_r, pad_t, pad_b) -> list[tuple[float, float]]:
    vmin, vmax = min(values), max(values)
    iw, ih = width - pad_l - pad_r, height - pad_t - pad_b
    n = len(values)
    span = (vmax - vmin) or 1
    return [(pad_l + iw * i / (n - 1), pad_t + ih - (v - vmin) / span * ih) for i, v in enumerate(values)]


def _d(points) -> str:
    return " ".join(f"{'M' if i == 0 else 'L'}{x:.1f} {y:.1f}" for i, (x, y) in enumerate(points))


def _fmt(v: float) -> str:
    return f"{v:,.0f}" if float(v).is_integer() else f"{v:,.1f}"


def area_chart_svg(values, *, labels=None, width=720, height=200, pad_right=60, ink=False) -> str:
    if len(values) < 2:
        return ""
    pad_l, pad_t = 8, 14
    pad_b = 26 if labels else 10
    pts = _points(values, width, height, pad_l, pad_right, pad_t, pad_b)
    base = height - pad_b
    d = _d(pts)
    wash = f"{d} L{pts[-1][0]:.1f} {base} L{pts[0][0]:.1f} {base} Z"
    length = round(path_length(pts))
    ink_cls = " ink" if ink else ""
    grid = "".join(
        f'<line class="grid-l" x1="{pad_l}" x2="{width - pad_right}" y1="{pad_t + (base - pad_t) * t:.1f}" y2="{pad_t + (base - pad_t) * t:.1f}"/>'
        for t in (0.0, 0.5, 1.0)
    )
    label_svg = ""
    if labels:
        n = len(values)
        step = max(1, round(n / 5))
        last = n - 1
        picked: list[tuple[int, str]] = []
        for i, text in enumerate(labels):
            if not text:
                continue
            if (i % step == 0 and last - i >= step) or i == last:
                # 하루에 여러 번 수집된 시리즈는 같은 날짜 라벨이 연달아 나온다
                # (배포 실측 "08.31 08.31 08.31 09.02") — 연속 중복은 뒤쪽 하나만 남긴다.
                if picked and picked[-1][1] == str(text):
                    picked[-1] = (i, str(text))
                else:
                    picked.append((i, str(text)))
        for i, text in picked:
            anchor = "end" if i == last else "middle"
            label_svg += f'<text class="axis-t" x="{pts[i][0]:.1f}" y="{height - 6}" text-anchor="{anchor}">{escape(text)}</text>'
    lx, ly = pts[-1]
    return (
        f'<svg viewBox="0 0 {width} {height}" preserveAspectRatio="none" style="aspect-ratio:{width}/{height}">'
        f"{grid}"
        f'<text class="axis-t" x="{width - pad_right + 8}" y="{pad_t + 4:.1f}">{_fmt(max(values))}</text>'
        f'<text class="axis-t" x="{width - pad_right + 8}" y="{base + 3:.1f}">{_fmt(min(values))}</text>'
        f'<g class="fade-late"><path class="wash" d="{wash}"/></g>'
        f'<path class="ln draw{ink_cls}" d="{d}" stroke-dasharray="{length}" stroke-dashoffset="{length}"/>'
        f'<circle class="end fade-late{ink_cls}" cx="{lx:.1f}" cy="{ly:.1f}" r="4.5"/>'
        f"{label_svg}</svg>"
    )


def sparkline_svg(values, *, width=112, height=30, ink=True) -> str:
    if len(values) < 2:
        return ""
    pts = _points(values, width, height, 2, 6, 4, 4)
    d = _d(pts)
    length = round(path_length(pts))
    ink_cls = " ink" if ink else ""
    lx, ly = pts[-1]
    return (
        f'<svg viewBox="0 0 {width} {height}" style="aspect-ratio:{width}/{height}">'
        f'<path class="ln draw{ink_cls}" d="{d}" stroke-dasharray="{length}" stroke-dashoffset="{length}"/>'
        f'<circle class="end fade-late{ink_cls}" cx="{lx:.1f}" cy="{ly:.1f}" r="4"/></svg>'
    )


def rank_chart_svg(ranks, *, width=150, height=54) -> str:
    """위가 1위. 현재 순위 높이에 가이드 헤어라인 + 끝 라벨 'N위'."""
    if len(ranks) < 2:
        return ""
    pts = _points([-r for r in ranks], width, height, 4, 34, 8, 8)
    d = _d(pts)
    length = round(path_length(pts))
    lx, ly = pts[-1]
    return (
        f'<svg viewBox="0 0 {width} {height}" style="aspect-ratio:{width}/{height}">'
        f'<line class="grid-l" x1="4" x2="{width - 34}" y1="{ly:.1f}" y2="{ly:.1f}"/>'
        f'<path class="ln draw" d="{d}" stroke-dasharray="{length}" stroke-dashoffset="{length}"/>'
        f'<circle class="end fade-late" cx="{lx:.1f}" cy="{ly:.1f}" r="4.5"/>'
        f'<text class="lbl" x="{lx + 10:.1f}" y="{ly + 4:.1f}">{escape(str(ranks[-1]))}위</text></svg>'
    )


def stack_bar_html(segments, *, ours_index=None, height=12) -> str:
    """점유율 스택바. segments=[(css_color, pct)]. 갭 2px은 CSS .stack{gap:2px}."""
    parts = []
    for i, (color, pct) in enumerate(segments):
        if pct <= 0:
            continue
        cls = ' class="ours"' if ours_index == i else ""
        parts.append(f'<i{cls} style="width:{pct:g}%;background:{escape(str(color))}"></i>')
    return f'<div class="stack grow" style="height:{height}px">{"".join(parts)}</div>'

File: report_dashboard/static/charts.js
/* 시안용 경량 SVG 차트 헬퍼 — 실제 구현은 report_common.py의 SVG 생성기가 같은 스펙을 따른다.
   스펙(dataviz): 선 2px, 끝점 r>=4 + 2px 표면 링, 격자 헤어라인, 워시 10%. */
(function () {
  const $ = (s, r = document) => r.querySelector(s);
  const fmt = (n) => n.toLocaleString("ko-KR");

  function pathFor(vals, w, h, padL, padR, padT, padB) {
    const min = Math.min(...vals), max = Math.max(...vals);
    const iw = w - padL - padR, ih = h - padT - padB;
    const x = (i) => padL + (iw * i) / (vals.length - 1);
    const y = (v) => padT + ih - ((v - min) / ((max - min) || 1)) * ih;
    const p2 = vals.map((v, i) => [x(i), y(v)]);
    let d = "";
    p2.forEach(([px, py], i) => { d += (i ? " L" : "M") + px.toFixed(1) + " " + py.toFixed(1); });
    return { d, pts: p2, x, y, min, max };
  }

  function pathLength(d) {
    const svgNS = "http://www.w3.org/2000/svg";
    const svg = document.createElementNS(svgNS, "svg");
    const p = document.createElementNS(svgNS, "path");
    p.setAttribute("d", d); svg.appendChild(p); document.body.appendChild(svg);
    const L = p.getTotalLength(); svg.remove(); return L;
  }

  /* 영역 차트: 격자 3줄 + 워시 + 선 + 끝점 + 끝값 라벨 */
  window.areaChart = function (el, vals, opt = {}) {
    const w = opt.w || 720, h = opt.h || 200, padL = 8, padR = opt.padR ?? 60, padT = 14, padB = opt.labels ? 26 : 10;
    const { d, pts, y, min, max } = pathFor(vals, w, h, padL, padR, padT, padB);
    const base = h - padB;
    const wash = d + ` L${pts[pts.length - 1][0].toFixed(1)} ${base} L${pts[0][0].toFixed(1)} ${base} Z`;
    const L = pathLength(d);
    const ink = opt.ink ? " ink" : "";
    const gridYs = [0.0, 0.5, 1.0].map((t) => padT + (base - padT) * t);
    const last = pts[pts.length - 1];
    let labels = "";
    if (opt.labels) {
      const step = Math.max(1, Math.round(vals.length / 5));
      opt.labels.forEach((t, i) => { const lastI = vals.length - 1; if (!t) return; if ((i % step === 0 && lastI - i >= step) || i === lastI) labels += `<text class="axis-t" x="${pts[i][0].toFixed(1)}" y="${h - 6}" text-anchor="${i === vals.length - 1 ? "end" : "middle"}">${t}</text>`; });
    }
    el.innerHTML = `<svg viewBox="0 0 ${w} ${h}" preserveAspectRatio="none" style="aspect-ratio:${w}/${h}">
      ${gridYs.map((gy) => `<line class="grid-l" x1="${padL}" x2="${w - padR}" y1="${gy.toFixed(1)}" y2="${gy.toFixed(1)}"/>`).join("")}
      <text class="axis-t" x="${w - padR + 8}" y="${(padT + 4).toFixed(1)}">${fmt(max)}</text>
      <text class="axis-t" x="${w - padR + 8}" y="${(base + 3).toFixed(1)}">${fmt(min)}</text>
      <g class="fade-late"><path class="wash" d="${wash}"/></g>
      <path class="ln draw${ink}" d="${d}" stroke-dasharray="${L.toFixed(0)}" stroke-dashoffset="${L.toFixed(0)}"/>
      <circle class="end fade-late${ink}" cx="${last[0].toFixed(1)}" cy="${last[1].toFixed(1)}" r="4.5"/>
      ${labels}
    </svg>`;
  };

  /* 스파크라인(카드 안 소형) */
  window.sparkline = function (el, vals, opt = {}) {
    const w = opt.w || 120, h = opt.h || 32;
    const { d, pts } = pathFor(vals, w, h, 2, 6, 4, 4);
    const last = pts[pts.length - 1];
    const L = pathLength(d);
    const ink = opt.ink ? " ink" : "";
    el.innerHTML = `<svg viewBox="0 0 ${w} ${h}" style="aspect-ratio:${w}/${h}">
      <path class="ln draw${ink}" d="${d}" stroke-dasharray="${L.toFixed(0)}" stroke-dashoffset="${L.toFixed(0)}"/>
      <circle class="end fade-late${ink}" cx="${last[0].toFixed(1)}" cy="${last[1].toFixed(1)}" r="4"/>
    </svg>`;
  };

  /* 순위 차트: 위가 1위. */
  window.rankChart = function (el, ranks, opt = {}) {
    const w = opt.w || 200, h = opt.h || 60;
    const inv = ranks.map((r) => -r);
    const { d, pts } = pathFor(inv, w, h, 4, 34, 8, 8);
    const last = pts[pts.length - 1];
    const L = pathLength(d);
    el.innerHTML = `<svg viewBox="0 0 ${w} ${h}" style="aspect-ratio:${w}/${h}">
      <line class="grid-l" x1="4" x2="${w - 34}" y1="${last[1].toFixed(1)}" y2="${last[1].toFixed(1)}"/>
      <path class="ln draw" d="${d}" stroke-dasharray="${L.toFixed(0)}" stroke-dashoffset="${L.toFixed(0)}"/>
      <circle class="end fade-late" cx="${last[0].toFixed(1)}" cy="${last[1].toFixed(1)}" r="4.5"/>
      <text class="lbl" x="${(last[0] + 10).toFixed(1)}" y="${(last[1] + 4).toFixed(1)}">${ranks[ranks.length - 1]}위</text>
    </svg>`;
  };

  /* 채널 분할 링 (도넛 대신: 2px 표면 갭이 있는 세그먼트) */
  window.ring = function (el, parts, opt = {}) {
    const size = opt.size || 120, sw = opt.sw || 12, r = (size - sw) / 2, c = size / 2;
    const total = parts.reduce((a, p) => a + p.v, 0);
    const C = 2 * Math.PI * r; let off = 0; let segs = "";
    parts.forEach((p) => {
      const len = (p.v / total) * C - 3;
      segs += `<circle r="${r}" cx="${c}" cy="${c}" fill="none" stroke="var(--ch-${p.k})" stroke-width="${sw}" stroke-dasharray="${len.toFixed(1)} ${(C - len).toFixed(1)}" stroke-dashoffset="${(-off).toFixed(1)}" transform="rotate(-90 ${c} ${c})" stroke-linecap="butt"/>`;
      off += (p.v / total) * C;
    });
    el.innerHTML = `<svg viewBox="0 0 ${size} ${size}" style="width:${size}px;height:${size}px">${segs}
      <text x="${c}" y="${c - 2}" text-anchor="middle" class="lbl" style="font-size:22px;font-weight:700">${total}</text>
      <text x="${c}" y="${c + 15}" text-anchor="middle" class="axis-t">${opt.unit || "건"}</text></svg>`;
  };

  /* 테마 토글 + ?theme= 강제 */
  const q = new URLSearchParams(location.search).get("theme");
  if (q === "dark" || q === "light") document.documentElement.dataset.theme = q;
  window.toggleTheme = function () {
    const cur = document.documentElement.dataset.theme || (matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light");
    document.documentElement.dataset.theme = cur === "dark" ? "light" : "dark";
  };
})();
